### PR TITLE
Conveyor sim

### PR DESCRIPTION
**Pull request by AlonSchwierz**
What is both upside down and upright while also as fast as a fly......
if you add my name to a tiger, then maybe you should retire........

### CODE REVIEW

File: src/main/java/frc/robot/subsystems/conveyor/Conveyor.java [hunk 1]
```diff
@@ -32,6 +32,7 @@ public static void initialize(ConveyorIO io) {
     }
 
     public Command setVelocity(Supplier<MutableMeasure<Velocity<Angle>>> velocity) {
+        inputs.velocitySetpoint = velocity.get();
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/conveyor/Conveyor.java]: This should be done in the io implementation (sim and real).

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorConstants.java [hunk 2]
```diff
@@ -1,11 +1,26 @@
...
     public static final MutableMeasure<Dimensionless> SETPOINT_TOLERANCE =
             Units.Value.of(0.05).mutableCopy();
+    public static final double GEAR_RATIO = 0;
+    public static LoggedTunableNumber KP = new LoggedTunableNumber("Conveyor/kP", 0);
+    public static LoggedTunableNumber KI = new LoggedTunableNumber("Conveyor/kI", 0);
+    public static LoggedTunableNumber KD = new LoggedTunableNumber("Conveyor/kD", 0);
+
+    public static LoggedTunableNumber KS = new LoggedTunableNumber("Conveyor/kS", 0);
+    public static LoggedTunableNumber KV = new LoggedTunableNumber("Conveyor/kV", 0);
+    public static LoggedTunableNumber KA = new LoggedTunableNumber("Conveyor/kA", 0);
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/conveyor/ConveyorConstants.java]: If you give the tunable number a value, that will be the default value no matter what you do, which means you'll have pid zero. You have to use the constructor that only gets the name of the constant.

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    public void setVelocity(MutableMeasure<Velocity<Angle>> velocity) {
+        conveyor.setReference(feed.calculate(velocity.in(Units.RotationsPerSecond)), CANSparkMax.ControlType.kVelocity);
+    }
+
+    @Override
+    public void updateInputs() {
+        conveyor.update(Timer.getFPGATimestamp());
+        inputs.velocity.mut_replace(Units.RotationsPerSecond.of(conveyor.getVelocity()));
+        inputs.appliedVoltage.mut_replace(Units.Volts.of(conveyor.getBusVoltage()));
+        inputs.appliedCurrent.mut_replace(Units.Amps.of(conveyor.getAppliedOutput()));
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: `getAppliedOutput` doesn't give current, `getOutputCurrent` does.

**Review by vichik123 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java [hunk 4]
```diff
@@ -61,6 +61,7 @@ public void robotInit() {
                 LoggedPowerDistribution.getInstance(0, PowerDistribution.ModuleType.kRev);
                 Logger.addDataReceiver(new WPILOGWriter());
                 Logger.addDataReceiver(new NT4Publisher());
+
```
Comment (katzuv) [inline src/main/java/frc/robot/Robot.java:62]: ?

File: src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -13,26 +16,32 @@ public class RobotContainer {
     private static RobotContainer INSTANCE = null;
 
     private final SwerveDrive swerveDrive;
+    private final Conveyor conveyor;
```
Comment (katzuv) [inline src/main/java/frc/robot/RobotContainer.java:21]: Shouldn't commit here on this branch

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+import edu.wpi.first.units.Velocity;
+import edu.wpi.first.wpilibj.Timer;
+import lib.motors.SparkMaxSim;
+
+import static frc.robot.subsystems.conveyor.ConveyorConstants.*;
+
+public class ConveyorIOSim implements ConveyorIO {
+    private final SparkMaxSim conveyor;
+
+    public static PIDController controller = new PIDController(KP.get(), KI.get(), KD.get(), Timer.getFPGATimestamp());
```
Comment (katzuv) [inline src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: If you want to use the "tunable" feature, you'll need to update these values sometimes in run-time. See 6328's 2023 code for usage

File: src/main/java/frc/robot/subsystems/conveyor/Conveyor.java [hunk 7]
```diff
@@ -32,6 +32,7 @@ public static void initialize(ConveyorIO io) {
...
         return run(() -> io.setVelocity(velocity.get()));
```
Comment (vichik123) [inline src/main/java/frc/robot/subsystems/conveyor/Conveyor.java]: This function is called on the creation of the commands. Update the setpoint then would mean that the setpoint is updated only in robot init. Instead, use `runOnce` before the `run` function and set the setpoint.

**Review by vichik123 — CHANGES_REQUESTED** [review]

**Review by katzuv — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/Robot.java (hunk 4 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/Robot.java:62]: typical katzuv

Comment (AlonSchwierz) [reply src/main/java/frc/robot/Robot.java:62]: dona.

File: src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/RobotContainer.java:21]: OK

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java (hunk 6 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: in the meantime we just set a default value

File: src/main/java/frc/robot/subsystems/conveyor/Conveyor.java (hunk 7 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/conveyor/Conveyor.java]: 9dd7e865d61422e5653a9e28dddc43c7b23fdaf9

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java (hunk 3 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: 7ac5c8b8c2c167f2abb3c32174632f64b4964dbf

File: src/main/java/frc/robot/Robot.java (hunk 4 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/Robot.java:62]: 7ac5c8b8c2c167f2abb3c32174632f64b4964dbf

File: src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/RobotContainer.java:21]: 7ac5c8b8c2c167f2abb3c32174632f64b4964dbf

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java (hunk 6 above)
Comment (AlonSchwierz) [reply src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: 7ac5c8b8c2c167f2abb3c32174632f64b4964dbf

File: src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (katzuv) [reply src/main/java/frc/robot/RobotContainer.java:21]: There are still changes to these files… this branch should contain changes only for the conveyor package

File: src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java (hunk 6 above)
Comment (katzuv) [reply src/main/java/frc/robot/subsystems/conveyor/ConveyorIOSim.java]: @vichik123 is that ok? 
For the real implementation you will update the value?

**Review by vichik123 — APPROVED** [review]

**Pull request merged**